---
title: "HiCRep analysis of 16p samples"
author: "Siddharth Reed"
date: "`r Sys.Date()`"
---
<style type="text/css">
.main-container {
  max-width: 1800px !important;
  margin-left: auto;
  margin-right: auto;
}
</style>

# Load Dependencies + HiCRep Data

```{r knitr}
library(knitr)
knitr::opts_chunk$set(
    results=FALSE,
    warning=FALSE,
    message=FALSE,
    dev=c('png', 'pdf', 'svg'),
    dpi=300
)
```
Load dependenies
```{r dependencies}
library(here)
here::i_am('notebooks/HiCRep.Rmd')
BASE_DIR <- here()
SCRIPT_DIR <- here('scripts')
source(file.path(BASE_DIR, 'scripts/locations.R'))
source(file.path(SCRIPT_DIR, 'utils.data.R'))
source(file.path(BASE_DIR, 'scripts/constants.R'))
source(file.path(SCRIPT_DIR, 'utils.plot.R'))
source(file.path(SCRIPT_DIR, 'hicrep', 'utils.hicrep.R'))
library(tidyverse)
library(magrittr)
```
Load hicrep data
```{r load_results}
hicrep.results <- 
    check_cached_results( 
        HICREP_RESULTS_FILE,
        # force_redo=TRUE,
        results_fnc=load_all_hicrep_results,
        samples_to_keep=NULL
    ) %>% 
    post_proces_hicrep_results() %>% 
    standardize_data_cols(skip.isMerged=TRUE)
# hicrep.results %>% head(2) %>% t()
# hicrep.results %>% dplyr::count(resolution, window.size, is.downsampled, chr)
```
plot objects
```{r plot_objects}
scc95_line <- 
    geom_hline(
        yintercept=0.95,
        color='black',
        linetype='solid',
        linewidth=0.15
    )
fill.gradient <- 
    scale_fill_gradient(
        low='grey90',
        high='blue'
    )
thin.top.colorbar <- 
    guides(
        fill=
            guide_colourbar(
                theme= 
                    theme(
                        legend.direction='horizontal',
                        legend.title=element_blank(),
                        legend.text=element_text(angle=30, hjust=1),
                        legend.key.height=unit(0.5, "lines"),
                        legend.key.width=unit(10, "lines")
                    )
            )
    )
```
How many different pairs of sample matrices were compared per batch
```{r n_samples, rows.print=5}
hicrep.results %>%
    filter(
        resolution == '10Kb',
        window.size == '5Mb',
        is.downsampled,
        chr == 'chr1'
    ) %>% 
    dplyr::count(
        isMerged,
        Edit,
        Genotype,
        name='n.sample.pairs'
    )
```

# Compare hyper-parameters differences {.tabset }

Compare values by hyper-param to showcase the variability
```{r hyperparam_boxplox, results='asis', fig.dim=c(10,6)}
hicrep.results %>% 
    make_nested_plot_tabs(
        group.cols=c('isMerged'),
        max.header.lvl=2,
        plot.fnc=plot_boxplot,
        x.var='window.size',
        x.scale.mode='discrete',
        y.var='hicrep.score',
        fill.var='is.downsampled.fct', 
        facet.row='resolution',
        facet.col='Celltype',
        scales='fixed',
        legend.position='top',
        y.axis.label.accuracy=0.01,
        axis.text.x=element_text(angle=35, hjust=1),
        plot.elements=list(scc95_line)
    )
```

# Plotting HiCRep scores {.tabset}

## Pick specific parameter set for downstream analyses

Pick a specific set of parameters to plto results for later
```{r pick_param_set, results='show'}
plot.df <- 
    hicrep.results %>% 
    # standardize sample order
    set_foldchange_direction_as_factor(
        sample_group_priority_fnc=SAMPLE_GROUP_PRIORITY_FNC,
        group1_colname='SampleID.P1',
        group2_colname='SampleID.P2'
    ) %>% 
    rename(
        'SampleID.P1'=Sample.Group.Numerator,
        'SampleID.P2'=Sample.Group.Denominator
    ) %>% 
    filter(
        is.downsampled,
        window.size == '5Mb'
        # resolution == '25Kb'
    )
plot.df %>% distinct(h, window.size, resolution, is.downsampled)
```

## Compare Scores across genotype comparisons {.tabset}

Plot score distributions between Genotypes
```{r compare_scores, results='asis', fig.dim=c(8,6)}
plot.df %>% 
    make_nested_plot_tabs(
        group.cols=c('isMerged'),
        max.header.lvl=3,
        plot.fnc=plot_boxplot,
        x.var='Genotype',
        x.scale.mode='discrete',
        y.var='hicrep.score',
        facet.row='resolution',
        facet.col='Celltype',
        legend.position='right',
        # y.scale.mode='',
        y.axis.label.accuracy=0.01,
        axis.text.x=element_text(angle=35, hjust=1),
        plot.elements=list(scc95_line)
    )
```

## HiCRep scores per chromosome  {.tabset }

```{r per_chr, results='asis', fig.dim=c(10,10)}
plot.df %>% 
    make_nested_plot_tabs(
        # group.cols=c('isMerged', 'resolution'),
        group.cols=c('isMerged', 'resolution'),
        max.header.lvl=3,
        plot.fnc=plot_boxplot,
        x.var='chr',
        x.scale.mode='discrete',
        y.var='hicrep.score',
        fill.var='Genotype',
        facet.row='Celltype',
        legend.position='right',
        y.axis.label.accuracy=0.01,
        axis.text.x=element_text(angle=35, hjust=1),
        plot.elements=list(scc95_line)
    )
```

## Heatmap {.tabset}

### Genome-Wide {.tabset}

```{r heatmap_scores_gw, results='asis', fig.dim=c(15,10)}
plot.df %>% 
    group_by(
        isMerged,
        resolution,
        SampleID.P1,
        SampleID.P2
    ) %>%
    summarize(
        # score.total=sum(hicrep.score),
        score.var=var(hicrep.score),
        score.min=min(hicrep.score),
        score.mean=mean(hicrep.score),
        score.median=median(hicrep.score)
    ) %>% 
    ungroup() %>% 
    pivot_longer(
        starts_with('score.'),
        names_to='metric',
        names_prefix='score.',
        values_to='score.stat'
    ) %>% 
    mutate(score.stat=round(score.stat, digits=2)) %>% 
    make_nested_plot_tabs(
        plot.fnc=plot_heatmap,
        group.cols=c('resolution', 'isMerged', 'metric'),
        max.header.lvl=4,
        x.var='SampleID.P1',
        y.var='SampleID.P2',
        fill.var='score.stat',
        label.var='score.stat',
        label.size=2.5,
        # facet.group='resolution',
        # facet.nrow=2,
        # facet.col='Genotype',
        legend.position='top',
        axis.text.x=element_text(angle=35, hjust=1),
        axis.title.x=element_blank(),
        axis.title.y=element_blank(),
        plot.elements=list(thin.top.colorbar, fill.gradient)
        # plot.elements=list(fill.gradient),
    )
```

### Per-Chr {.tabset}

```{r heatmap_scores_chr, results='asis', fig.dim=c(15,15)}
plot.df %>% 
    mutate(hicrep.score=round(hicrep.score, digits=2)) %>% 
    make_nested_plot_tabs(
        plot.fnc=plot_heatmap,
        group.cols=c('isMerged', 'resolution'),
        max.header.lvl=4,
        x.var='SampleID.P1',
        y.var='SampleID.P2',
        fill.var='hicrep.score',
        facet.group='chr',
        facet.ncol=5,
        # label.var='hicrep.score',
        # facet.col='Genotype',
        legend.position='top',
        axis.title.x=element_blank(),
        axis.title.y=element_blank(),
        axis.text.x=element_text(angle=45, hjust=1),
        plot.elements=list(thin.top.colorbar, fill.gradient)
    )
```

