---
title: "HiCRep analysis of 16p samples"
author: "Siddharth Reed"
date: "`r Sys.Date()`"
output: 
    html_document:
        keep_md: yes
        toc: true
        toc_float: true
        theme: paper
        code_folding: hide
        df_print: paged
---
<style type="text/css">
.main-container {
  max-width: 1800px !important;
  margin-left: auto;
  margin-right: auto;
}
</style>

# Prepare data

```{r knitr}
library(knitr)
knitr::opts_chunk$set(
    dev=c('png', 'pdf', 'svg'),
    dpi=300
)
```
Set up file locations + load dependenies
```{r file_locs, results=FALSE, message=FALSE, warning=FALSE}
library(here)
here::i_am('notebooks/hicrep.Rmd')
BASE_DIR <- here()
SCRIPT_DIR <- here('scripts')
source(file.path(SCRIPT_DIR, 'locations.R'))
source(file.path(SCRIPT_DIR, 'constants.R'))
source(file.path(SCRIPT_DIR, 'utils.data.R'))
source(file.path(SCRIPT_DIR, 'utils.plot.R'))
source(file.path(SCRIPT_DIR, 'utils.hicrep.R'))
library(tidyverse)
library(magrittr)
library(ggplot2)
library(ggh4x)
library(purrr)
```
Load hicrep data
```{r load_results, results=FALSE, message=FALSE, warning=FALSE}
sample.metadata <- 
    load_sample_metadata() %>%
    mutate(Sequencing.Run=factor(Sequencing.Run)) %>%
    mutate(Batch=glue('Batch {as.integer(Sequencing.Run)}'))
hicrep.results <- 
    check_cached_results( 
        HICREP_RESULTS_FILE,
        force_redo=FALSE,
        return_data=TRUE,
        results_fnc=load_all_hicrep_results
    ) %>%
    left_join(
        RESOLUTION_IDEAL_H,
        relationship='many-to-many',
        by=join_by(resolution)
    ) %>% 
    # get_min_resolution_per_matrix(filter_res=FALSE) %>% 
    filter(ReadFilter == 'mapq_30 vs mapq_30') %>%
    select(-c(ReadFilter)) %>% 
    filter(!grepl('iN', Celltype))
# hicrep.results
```
Print sample rows
```{r example_results, rows.print=5}
hicrep.results %>% 
    arrange(
        across(
            c(
                resolution,
                h,
                is.downsampled,
                window.size,
                chr, 
                SampleNumber
            )
        )
    )
```

# Params Used

How many different parameter combinations we tried
```{r param_combos, rows.print=5}
hicrep.results %>%
    count(
        resolution,
        h,
        is.downsampled,
        window.size,
        chr
    )
```

How many different pairs of sample matrices were compared
```{r n_samples, rows.print=5}
hicrep.results %>%
    group_by(
        across(
            -c(
                chr,
                resolution,
                h,
                is.downsampled,
                window.size
            )
        )
    ) %>% 
    count()
```

Prepare data for plotting
```{r prep_plot_df, include=FALSE}
plot.df <- 
    hicrep.results %>%
    mutate(
        chr=factor(chr, levels=CHROMOSOMES),
        across(
            c(resolution, window.size),
            ~ factor(
                scale_numbers(as.integer(.x)),
                levels=scale_numbers(sort(unique(as.integer(.x))))
            )
        )
    ) %>% 
    # Add batch information
    separate_wider_delim(
        Sample.ID,
        delim=' vs ',
        names=c('Sample.ID.A', 'Sample.ID.B')
    ) %>% 
    left_join(
        sample.metadata %>% select(Sample.ID, Batch),
        by=join_by(Sample.ID.A == Sample.ID)
    ) %>% 
    left_join(
        sample.metadata %>% select(Sample.ID, Batch),
        suffix=c('.A', '.B'),
        by=join_by(Sample.ID.B == Sample.ID)
    ) %>% 
    mutate(
        across(
            starts_with('Batch'),
            ~ ifelse(is.na(.x), 'Merged', .x)
        ),
        Batch=
            pmap_chr(
                list(Batch.A, Batch.B),
                ~ paste(sort(c(...)), collapse=' vs ')
            )
    ) %>% 
    select(-ends_with(c('.A', '.B')))
```

# Compare hyper-params

Compare values by hyper-param to showcase the variability
```{r hyperparam_boxplox, results='asis', echo=FALSE, out.width='100%', fig.height=7}
plot.df %>% 
    make_nested_plot_tabs(
        group_cols=c('isMerged'),
        plot_fnc=plot_hicrep_boxplot,
        sample_group='window.size',
        fill_var='is.downsampled', 
        facet_col='resolution',
        facet_row='Batch',
        max_header_lvl=3
    )
```

```{r pick_param_set}
plot.df <- 
    plot.df %>% 
    filter(
        is.downsampled,
        window.size == '5Mb',
        h == Ideal_H,
        resolution == '25Kb'
    )
plot.df %>% distinct(h, window.size, resolution, is.downsampled)
```

# Lab Meeting Plot

```{r lab_meeting, results='asis', fig.align='center', out.width='100%', fig.height=6}
plot.df %>% 
    make_nested_plot_tabs(
        group_cols=c('isMerged'),
        fill_var=NA,
        plot_fnc=plot_hicrep_boxplot,
        sample_group='chr',
        facet_col='Genotype'
    )
```

# Plotting HiCRep scores {.tabset}

## Compare Individual Samples

Boxplot of hicrep scores by various levels
```{r new_boxplot, echo=FALSE, fig.align='center', out.width='100%', fig.height=7}
plot.df %>% 
    filter(isMerged == 'Individual vs Individual') %>% 
    plot_hicrep_boxplot(
        sample_group='Genotype',
        facet_row='resolution',
        facet_col='Celltype',
        mark_df={. %>% filter(chr %in% c('chr16'))},
        mark_size=0.85
    )
```

## Comapre Merged Samples

Boxplot of hicrep scores by various levels
```{r merged_boxplot, echo=FALSE, fig.align='center', out.width='100%', fig.height=7}
plot.df %>% 
    filter(isMerged == 'Merged vs Merged') %>% 
    plot_hicrep_boxplot(
        sample_group='Genotype',
        facet_col='Celltype',
    )
```

## HiCRep scores per chromosome 

See if there are any trends in HiCRep scores across chromosomes and if they are influenced by HiCRep params or samples genotype
```{r per_chr, results='asis', echo=FALSE, out.width='100%', fig.height=7, eval=FALSE}
plot.df %>%
    filter(isMerged == 'Individual vs Individual') %>% 
    filter(is.downsampled) %>% 
    make_nested_plot_tabs(
        group_cols=c('resolution', 'Celltype'),
        plot_fnc=plot_hicrep_boxplot,
        sample_group='chr',
        # fill_var='is.downsampled', 
        fill_var='window.size', 
        facet_row='Genotype',
        yintercept=0.95,
        max_header_lvl=3
    )
```
