---
title: "Loop Integration Analysis for 16p"
author: "Siddharth Reed"
date: "`r Sys.Date()`"
---
<style type="text/css">
.main-container {
  max-width: 1800px !important;
  margin-left: auto;
  margin-right: auto;
}
</style>

# Dependencies + Load Data

```{r load_data, eval=FALSE, cache=TRUE}
regions.df <- 
    GENOMIC_REGIONS %>% 
    dplyr::rename_with(~ str_remove(.x, 'region.')) %>%
    select(region, chr, start, end)
ccres.df <- load_encode_ccres()
degs.df <- load_DESeq2_results()
trade.df <- load_all_TRADE_analysis()
trade.df %>%
    select(-c(Celltype, Genotype, TRADE.sig.genes)) %>%
    dplyr::rename('adj.method'=TRADE.adjust.method) %>% 
    pivot_longer(starts_with('TRADE'), names_to='stat', names_prefix='TRADE.', values_to='value') %>%
    pivot_wider(names_from=adj.method, values_from=value) %>%
    # filter((stat %in% c('TWI', 'log.likelyhood'))) %>%
    filter(!(stat %in% ('num.nonsig', 'TWI', 'log.likelyhood'))) %>%
    print(n=Inf)

idr2d.df <- 
    FILTERED_LOOPS_FILTERED_IDR2D_RESULTS_FILE %>% 
    read_tsv(show_col_types=FALSE) %>% 
    post_process_IDR2D_results() %>% 
    standardize_data_cols()
```
```{r knitr}
library(knitr)
knitr::opts_chunk$set(
    results=FALSE,
    message=FALSE,
    warning=FALSE,
    dev=c('png', 'pdf', 'svg'),
    # dev=c('pdf', 'png'),
    dpi=300
)
```
Set up file locations + load dependences
```{r dependencies}
library(here)
here::i_am('notebooks/Loop.Integration.Rmd')
BASE_DIR <- here()
source(file.path(BASE_DIR, 'scripts/locations.R'))
source(file.path(SCRIPT_DIR, 'utils.data.R'))
source(file.path(BASE_DIR, 'scripts/constants.R'))
source(file.path(SCRIPT_DIR, 'utils.plot.R'))
source(file.path(SCRIPT_DIR, 'utils.annotations.R'))
source(file.path(SCRIPT_DIR, 'loops/utils.loops.R'))
library(tidyverse)
library(magrittr)
```
Load cCRE annotation data
```{r load_ccre_loop_data, eval=TRUE, cache=TRUE}
library(furrr)
plan(multisession, workers=length(availableWorkers()))
n.ccres.df <- 
    check_cached_results(
        results_file=LOOP_IDR2D_CCRE_MAPPING_FILE,
        # force_redo=TRUE,
        results_fnc=count_ccres_per_loop
    )
```
Load DEG results data
```{r load_deg_loop_data, eval=FALSE, cache=TRUE}
# Join annotations and loops
degs.loops.df <- 
    # Load IDR2D results for all comparisons of sample groups 
    check_cached_results(
        results_file=LOOPS_IDR2D_RESULTS_FILE,
        # force_redo=TRUE,
        results_fnc=load_all_IDR2D_results
    ) %>% 
    # filter(max.gap.bins == 5) %>% 
    post_process_IDR2D_results() %>% 
    standardize_data_cols() %>% 
    # Join cCRES to loops
    left_join(
        load_DESeq2_results(),
        relationship='many-to-many',
        by=
            join_by(
                chr,
                within(
                    y$start, y$end,   
                    x$anchor.left, x$anchor.right
                )
            )
    )
```

# Compare loops to cCRES

## Genome-Wide

```{r plot_ccres_vs_loop_reproducibility_barplot_gw, results='asis', fig.dim=c(7,7)}
# n.ccre.loops.df
n.ccres.df
n.ccres.df %>% count(resolution, max.gap.bins, comparison, chr, region, cCRE.type)
n.ccres.df %>% 
    filter(region == chr) %>% 
    group_by(resolution, max.gap.bins, comparison, chr, region, cCRE.type) %>%
    summarize(n=sum(n), total.cCRE=unique(total.cCREs))
    count(resolution, max.gap.bins, comparison, chr, region, cCRE.type)
```

## Per-Chr

```{r plot_ccres_vs_loop_reproducibility_barplot_chr, results='asis', fig.dim=c(7,7)}
```

# Plot expression of genes between loops  {.tabset}

```{r plot_expr_of_loops, results='asis', fig.dim=c(10,6), eval=FALSE}
loop.expr.df %>% 
    filter(weight == 'balanced', resolution == '10Kb', kernel == 'donut') %>% 
    filter(TPM.mean < 200) %>% 
    filter(TPM.mean > 50) %>% 
    make_nested_plot_tabs(
        max.header.lvl=3,
        group.cols=c('weight', 'resolution', 'comparison'),
        plot.fnc=plot_boxplot,
        # x.var='chr',
        x.var='is.loop.shared',
        x.scale.mode='discrete',
        y.var='TPM.mean',
        fill.var='SampleID',
        # fill.var='SampleID',
        fill.scale.mode='discrete',
        # label.var='',
        # facet.group='chr',
        # facet.ncol=6,
        # facet.row='is.loop.shared',
        # facet.col='is.loop.shared',
        scales='free_y',
        legend.position='right',
        # fill.scale.mode='m',
        strip.text=element_text(size=10),
        axis.text.x=element_text(angle=45, hjust=1),
        axis.title.x=element_blank(),
        axis.title.y=element_blank()
    )
```

# Compare Loops across Genotypes {.tabset}

## Plot anchor differences boxplot {.tabset}

```{r plot_compare_anchors_boxplot, results='asis', fig.dim=c(10,6), eval=FALSE}
diff.df %>% 
    # filter(!metric %in% c('start', 'end', 'length')) %>% 
    filter(metric %in% c('length', 'enrichment', 'log10_qval')) %>% 
    mutate(diff=abs(diff)) %>% 
    make_nested_plot_tabs(
        group.cols=c('weight', 'resolution'),
        max.header.lvl=3,
        plot.fnc=plot_boxplot,
        x.var='chr',
        y.var='diff',
        fill.var='SampleID',
        facet.row='metric',
        x.scale.mode='discrete',
        scales='free_y',
        legend.position='top',
        axis.title.x=element_blank(),
        strip.text=element_text(size=20),
        axis.text.x=element_text(angle=45, hjust=1)
    )
```

## Plot anchor differences scatter {.tabset}

```{r plot_compare_anchors_scatter, results='asis', fig.dim=c(10,6), eval=FALSE}
diff.df %>% 
    # filter(metric %in% c(start, end, length)) %>% 
    # filter(!metric %in% c('start', 'end', 'length')) %>% 
    filter(metric %in% c('enrichment', 'log10_qval')) %>% 
    make_nested_plot_tabs(
        group.cols=c('weight', 'resolution'),
        max.header.lvl=3,
        plot.fnc=plot_jitter,
        x.var='P1',
        y.var='P2',
        color.var='SampleID',
        # facet.row='metric',
        facet.group='metric',
        facet.ncol=1,
        scales='free',
        size=0.75,
        legend.position='top',
        axis.title.x=element_blank(),
        axis.title.y=element_blank(),
        strip.text=element_text(size=20),
        axis.text.x=element_text(angle=45, hjust=1)
    )
```

## Plot Anchor MoCs {.tabset}

```{r plot_compare_anchors_MoC, results='asis', fig.dim=c(10,6), eval=FALSE}
similarity.df %>% 
    # select(-c(ends_with('diff'))) %>% 
    group_by(weight, resolution, chr, kernel, SampleID) %>%
    summarize(MoC=(sum(moc.inner) - 1) * unique(MoC.norm)) %>% 
    ungroup() %>% 
    make_nested_plot_tabs(
        # group.cols=c('type', 'isMerged', 'weight', 'resolution'),
        group.cols=c('weight', 'resolution'),
        max.header.lvl=3,
        plot.fnc=plot_barplot,
        x.var='chr',
        y.var='MoC',
        # color.var='Genotype',
        # shape.var='Genotype',
        # x.var='Genotype',
        fill.var='SampleID',
        # facet.col='chr',
        # facet.row='context',
        x.scale.mode='discrete',
        # y.scale.mode='mb',
        # scales='free_y',
        # size=0.25,
        legend.position='top',
        axis.title.x=element_blank(),
        axis.text.x=element_text(angle=45, hjust=1)
    )
```

