---
title: "TAD Analysis for 16p Samples"
author: "Siddharth Reed"
date: "`r Sys.Date()`"
---
<style type="text/css">
.main-container {
  max-width: 1800px !important;
  margin-left: auto;
  margin-right: auto;
}
</style>

# Set Up

```{r knitr}
options(scipen=999)
options(dplyr.summarise.inform=FALSE)
library(knitr)
knitr::opts_chunk$set(
    results=FALSE,
    warning=FALSE, 
    message=FALSE,
    dev=c('svg', 'png', 'pdf'),
    dpi=300
)
```
Set up file locations + load dependenies
```{r dependencies}
library(here)
here::i_am('notebooks/TADs.Rmd')
BASE_DIR <- here()
source(file.path(BASE_DIR, 'scripts', 'locations.R'))
source(file.path(SCRIPT_DIR, 'utils.data.R'))
source(file.path(BASE_DIR, 'scripts', 'constants.R'))
source(file.path(SCRIPT_DIR, 'utils.plot.R'))
source(file.path(SCRIPT_DIR, 'TADs', 'utils.TADs.R'))
source(file.path(SCRIPT_DIR, 'TADs', 'utils.TADCompare.R'))
library(tidyverse)
library(magrittr)
```

# Load TAD results {.tabset}

Load all hiTAD annotations
```{r load_TADs}
TADs.df <- 
    check_cached_results(
        results_file=ALL_TAD_RESULTS_FILE,
        # force_redo=TRUE,
        results_fnc=
            function(){
                load_all_TAD_results_for_TADCompare() %>% 
                filter(TAD.method != 'spectralTAD') %>% 
                dplyr::rename('boundaries'=pre_tads) %>%
                select(-c(chr)) %>% 
                unnest(boundaries)
        }
    ) %>% 
    separate_wider_delim(
        SampleID,
        delim='.',
        names=c(NA, 'Celltype', 'Genotype', NA, NA),
        cols_remove=FALSE
    ) %>%
    filter(resolution == 50000) %>% 
    standardize_data_cols()
```
define the most similar pairs of TADs between all between-sample comparisons
```{r calcualte_TAD_similarities}
# library(furrr)
# plan(multisession, workers=length(availableWorkers()))
all.TAD.pairs.df <- 
    TADs.df %>% 
    select(-c(Celltype, Genotype, SampleID)) %>% 
    nest(boundaries=c(start, end, length)) %>% 
    check_cached_results(
        results_file=ALL_TAD_SIMILARITY_RESULTS_FILE,
        # force_redo=TRUE,
        results_fnc=calculate_all_boundary_similarities,
        boundaries.df=.,
        only_keep_overlaps=TRUE,
        sample.group.comparisons=
            ALL_SAMPLE_GROUP_COMPARISONS %>%
            rename(
                'Sample.Group.P1'=Sample.Group.Numerator,
                'Sample.Group.P2'=Sample.Group.Denominator,
            ),
        pair_grouping_cols=
            c(
                # 'isMerged',
                # 'weight',
                'resolution',
                'TAD.method',
                'TAD.params',
                'chr'
            )
    ) %>% 
    mutate(chr=factor(chr, levels=CHROMOSOMES))
# plan(sequential)
# similarities.df
```

# Number of TADs {.tabset}

## Genome-Wide {.tabset}

```{r plot_TAD_n_gw, results='asis', fig.dim=c(10,6)}
TADs.df %>% 
    dplyr::count(
        # method, TAD.params, resolution, Sample.Group, chr,
        across(-c('chr', 'start', 'end', 'length')),
        name='nTADs'
    ) %>% 
    make_nested_plot_tabs(
        group.cols=c('TAD.method', 'resolution'),
        max.header.lvl=3,
        plot.fnc=plot_barplot,
        x.var='Sample.Group',
        y.var='nTADs',
        fill.var='Genotype',
        facet.col='Celltype',
        scales='free_x',
        legend.position='right',
        # legend.text=element_text(angle=45, hjust=1),
        axis.title.x=element_blank(),
        axis.title.y=element_blank(),
        axis.text.x=element_text(angle=45, hjust=1)
    )
```

## Per Chr {.tabset}

```{r plot_TAD_n_chr, results='asis', fig.dim=c(10,6)}
TADs.df %>% 
    dplyr::count(
        # method, TAD.params, resolution, Sample.Group, chr,
        across(-c('start', 'end', 'length')),
        name='nTADs'
    ) %>% 
    make_nested_plot_tabs(
        group.cols=c('TAD.method', 'resolution'),
        max.header.lvl=3,
        plot.fnc=plot_barplot,
        x.var='chr',
        y.var='nTADs',
        fill.var='Sample.Group',
        # label.var='nTADs',
        # label.size=2,
        facet.row='Celltype',
        scales='free_y',
        # legend.position='top',
        # legend.text=element_text(angle=45, hjust=1),
        axis.title.x=element_blank(),
        axis.title.y=element_blank(),
        axis.text.x=element_text(angle=45, hjust=1)
    )
```

# TAD Lengths {.tabset }

```{r plot_TAD_lengths, results='asis', fig.dim=c(8,6)}
TADs.df %>% 
    make_nested_plot_tabs(
        # group.cols=c('isMerged', 'weight', 'resolution'),
        # group.cols=c('method', 'resolution'),
        group.cols=c('TAD.method', 'resolution'),
        max.header.lvl=2,
        plot.fnc=plot_boxplot,
        x.var='chr',
        y.var='length',
        fill.var='Sample.Group',
        facet.row='Celltype',
        scales='free_y',
        y.scale.mode='mb',
        axis.text.x=element_text(angle=45, hjust=1)
    )
```

# TAD Similarity (MoC) {.tabset }

Calculate MoC for all pairs of TAD annotation sets
```{r plot_MoCs_boxplot, results='asis', fig.dim=c(10, 6)}
all.TAD.pairs.df %>% 
    # mutate(MoC=moc.inner / MoC.norm) %>% 
    filter(stat == 'MoC') %>% 
    pivot_wider(names_from='stat', values_from='value') %>% 
    mutate(comparison=glue('{Sample.Group.P2} vs {Sample.Group.P1}')) %>% 
    make_nested_plot_tabs(
        # group.cols=c('isMerged', 'weight', 'resolution'),
        # group.cols=c('method', 'TAD.params', 'resolution'),
        # group.cols=c('method', 'resolution'),
        group.cols=c('TAD.method', 'resolution'),
        max.header.lvl=2,
        plot.fnc=plot_barplot,
        x.var='chr',
        y.var='MoC',
        fill.var='comparison',
        # facet.row='Genotype',
        facet.row='Celltype',
        legend.position='right',
        axis.text.x=element_text(angle=45, hjust=1),
        axis.title.x=element_blank(),
        axis.title.y=element_blank(),
        x.axis.label.accuracy=0.01
    )
```

# Boundary Differences {.tabset }

Plot MoC of all TADs that overlap between conditions
```{r plot_MoCs_heatmap, results='asis', eval=FALSE}
all.TAD.pairs.df %>% 
    filter(!context %in% c('jaccard', 'boundary')) %>% 
    # filter(!stat %in% c('maximum', 'minimum')) %>% 
    filter(stat %in% c('mean', 'maximum')) %>% 
    make_nested_plot_tabs(
        # group.cols=c('isMerged', 'weight', 'resolution'),
        # group.cols=c('method', 'TAD.params', 'resolution', 'stat'),
        group.cols=c('method', 'resolution', 'stat'),
        max.header.lvl=2,
        plot.fnc=plot_barplot,
        x.var='chr',
        y.var='value',
        fill.var='Sample.Group',
        # shape.var='stat',
        facet.row='context',
        x.scale.mode='discrete',
        y.scale.mode='mb',
        scales='free_y',
        legend.position='top',
        size=1.3,
        axis.title.x=element_blank(),
        axis.title.y=element_blank(),
        axis.text.x=element_text(angle=45, hjust=1)
    )
```

