# Dependencies
# library(tidyverse)
# library(glue)
# library(purrr)
###############
# Load various QC data files/sets of files
load_pairtools_stats <- function(
    stats_file_suffix='.HiC.hg38.dedup.stats',
    ...){
    # Load all stats
    all.stats.df <- 
        PAIRS_DIR  %>%
        list.files(
            full.names=FALSE,
            recursive=TRUE,
            pattern=glue('*{stats_file_suffix}')
        ) %>%
        tibble(fileinfo=.) %>% 
        mutate(filepath=file.path(PAIRS_DIR, fileinfo)) %>% 
        mutate(Sample.ID=str_remove(basename(fileinfo), stats_file_suffix)) %>% 
        select(-c(fileinfo)) %>% 
        mutate(
            stats=
                purrr::pmap(
                    .l=.,
                    function(filepath, ...){
                        filepath %>% 
                        read_tsv(
                            show_col_types=FALSE,
                            col_names=c('stat', 'value')
                        )
                    }
                )
        ) %>%
        select(-c(filepath)) %>%
        unnest(stats)
    # Get total number of contacts per sample
    total.contacts.df <- 
        all.stats.df %>% 
        filter(!grepl('/', stat)) %>%
        filter(stat == 'total_nodups') %>% 
        select(Sample.ID, value) %>%
        dplyr::rename('total.unique.contacts'=value)
    # Matrix-wide stats
    general.stats.df <- 
        all.stats.df %>% 
        filter(!grepl('/', stat)) %>%
        mutate(
            category=
                case_when(
                    grepl('total', stat) ~ 'Reads', 
                    TRUE ~ 'Contacts'
                )
        ) %>% 
        left_join(
            total.contacts.df,
            by='Sample.ID'
        )
    # stats about contact distances
    dist.stats.df <- 
        all.stats.df %>% 
        filter(grepl('^dist_freq', stat)) %>%
        separate_wider_delim(
            stat,
            delim='/',
            names=
                c(
                    'interaction',
                    'range',
                    'orientation'
                )
        ) %>%
        mutate(range=str_remove(range, fixed('+'))) %>% 
        separate_wider_delim(
            range,
            delim='-',
            names=c('range.start', 'range.end'),
            too_few='align_start'
        ) %>%
        mutate(across(starts_with('range'), ~ as.integer(.x))) %>% 
        mutate(interaction='cis') %>%
        left_join(total.contacts.df, by='Sample.ID')
    # Contact Frequency Stats per Chromosome
    chr.stats.df <- 
        all.stats.df %>% 
        filter(grepl('^chrom_freq', stat)) %>%
        separate_wider_delim(
            stat,
            delim='/',
            names=
                c(
                    'stat',
                    'chr1',
                    'chr2'
                )
        ) %>% 
        left_join(total.contacts.df, by='Sample.ID')
    # Return named list of all stuff
    return(
        list(
            'general.stats'=general.stats.df,
            'distance.stats'=dist.stats.df,
            'chr.stats'=chr.stats.df
        )
    )
}

load_genome_coverage <- function(
    delim='-',
    weights=NULL,
    resolutions=NULL){
    # input_dir=COVERAGE_DIR; suffix='coverage.tsv'; param_names=c('weight', 'resolution'); delim='-'; resolutions=c(1000, 100000); weights=c('raw')
    # List input files (generated by cooltools coverage)
    parse_results_filelist(
        input_dir=COVERAGE_DIR,
        suffix='-coverage.tsv',
        filename.column.name='matrix.name',
        param_delim='_',
    ) %>%
    process_matrix_name() %>% 
    # Ignore results not matching param values
    { 
        if (is.null(resolutions)) { 
            . 
        } else { 
            filter(., resolution %in% resolutions) 
        } 
    } %>%
    { 
        if (is.null(weights)) { 
            . 
        } else { 
            filter(., weight %in% weights) 
        } 
    } %>%
    mutate(
        coverage=
            purrr::pmap(
                .l=.,
                .f=function(filepath, ...){
                    read_tsv(
                        filepath,
                        show_col_types=FALSE,
                        progress=FALSE
                    )
                }
            )
    ) %>%
    unnest(coverage) %>%
    rename_with(~ str_remove(.x, '_(raw|weight)')) %>% 
    rename(
        'chr'=chrom,
        'coverage.cis'=cov_cis,
        'coverage.total'=cov_tot,
    ) %>% 
    mutate(chr=factor(chr, levels=CHROMOSOMES)) %>%  
    pivot_longer(
        starts_with('coverage'),
        names_to='metric',
        names_prefix='coverage.',
        values_to='coverage'
    )
}

# Plot Matrix QC stuff
plot_qc_barplot <- function(
    plot.df,
    y_val,
    facet_row,
    facet_col,
    ...){
    plot.df %>%
    ggplot(
        aes(
            x=Category,
            y=.data[[y_val]],
            fill=Sample.ID
        )
    ) +
    geom_col(position = "dodge") +
    # scale_y_continuous(breaks=~ seq(.x, .y, 5)) +
    scale_y_continuous(
        breaks=seq(0, max(plot.df[[y_val]]), 5),
        expand=c(0, 0)
    ) +
    facet_grid2(
        rows=vars(!!sym(facet_row)),
        cols=vars(!!sym(facet_col)),
        scales='fixed'
    ) +
    labs(y=y_val) +
    theme(
        axis.title.x=element_blank(),
        axis.text.x=
            element_text(
                hjust=1,
                angle=35
            )
    ) + 
    add_ggtheme()
}

    ) +
    add_ggtheme()
}

plot_cistrans_chromosome_heatmap <- function(
    plot.df,
    fill_var,
    ...){
    plot.df %>%
    ggplot(
        aes(
            x=chr1,
            y=chr2
        )
    ) +
    geom_tile(
        aes(
            fill=.data[[fill_var]]
        )
    ) +
    scale_fill_gradient(
        low='grey90',
        high='red'
    ) +
    facet_wrap(~ Sample.ID) +
    theme(
        legend.position='top',
        axis.text.x=element_text(angle=45, hjust=1)
    ) +
    add_ggtheme()
}
